### PR TITLE
Refactor gunmods

### PR DESCRIPTION
**Pull request by mugling**
Nobody likes the gunmod code. This PR will try and improve the more ugly aspects and provide some minimal documentation.

### CODE REVIEW

**Comment by mugling** [comment]
Commits are hopefully self explanatory.

I'm struggling with documentation of three fields in `JSON_INFO.md`.

* Can someone explain how `sight_dispersion` and `aim_speed` work?
* Is specifying `skill` for gunmods actually necessary?

**Comment by Coolthulhu** [comment]
You could make the jsons shorter by one pair of square brackets by special casing the 2 element array if its first element is a string and not an array.

> Is specifying skill for gunmods actually necessary?

Currently not at all, they all use the same skill.
Could be useful in the future, where complex mods might require mechanics or even electronics.
> Can someone explain how sight_dispersion and aim_speed work?

`sight_dispersion` prevents a given sight from being considered if current recoil is above that.
`aim_speed` slows down aiming with a given sight.
When aiming, you always use the sight with the lowest `aim_speed` that has `sight_dispersion` below current recoil.

**Comment by mugling** [comment]
> You could make the jsons shorter by one pair of square brackets by special casing the 2 element array if its first element is a string and not an array.

Sorry you lost me there

**Comment by mugling** [comment]
Done for now. I'm going to come back to this in a further PR which will probably also include refactoring and reaming of `aim_speed` to `aim_cost` as per #15144

**Comment by Coolthulhu** [comment]
I thought you mean "skill" as in the one used to install. I'm not too sure about skill used to fire it. Did you test it? Does it work with the skill being specified only in the gun data?

File: src/item.cpp [hunk 1]
```diff
@@ -3663,6 +3614,23 @@ skill_id item::gun_skill() const
     return type->gun->skill_used;
 }
 
+std::string item::gun_type() const
```
Comment (Coolthulhu) [inline src/item.cpp]: What is it for? Why special case crossbow and bow?

**Comment by mugling** [comment]
> I'm not too sure about skill used to fire it.

The code never checks for this

File: src/item.cpp (hunk 1 above)
Comment (mugling) [inline src/item.cpp]: Returns the field used for `mod_targets`

**Comment by BevapDin** [comment]
If you change the JSON format, you should update all the affected JSON files, I see "skill_required" being used in data/mods/FictonalWeapons/fic_mods.json, which is not updated here. You can use `./cataclysm --check-mods` to check all mods. It loads the data of each mod and runs the consistency checks on them.

**Comment by mugling** [comment]
> You can use ./cataclysm --check-mods to check all mods

That I did not know

**Comment by mugling** [comment]
Done

**Comment by Coolthulhu** [comment]
Significant UI regression: `can_use` only prints "you lack the skill in marksmanship to use x". The required skill level can't be learned from in-game UI alone.

**Comment by mugling** [comment]
@Coolthulhu I've gone with the more generic approach of showing stat/skill requirements in `item::info` as `min_skills` can apply to more than just gunmods. It requires no further i18n efforts.

e67c569 also extends `item::gunmod_compatible` to allow ignoring of status effects (loaded, jammed etc) so that we can still potentially compatible items in the filtered inventory but then later refuse with the appropriate message when selected by the player.

**Comment by mugling** [comment]
I've snuck in a fix for #15354 as it's a very appropriate use for `item::gunmods`. Which other PR does this conflict with?

**Comment by Coolthulhu** [comment]
Minor bug, probably not caused here: setting gun mode to bayonet and removing the bayonet will keep displaying the mode as bayonet until different mode is selected. Doesn't allow any exploits, though.

**Comment by Coolthulhu** [comment]
Another minor gunmod-related bug: burst size can get to 0, but this doesn't prevent setting gun mode to burst.

File: src/item.cpp [hunk 2]
```diff
@@ -5821,12 +5789,10 @@ itype *item::find_type( const itype_id &type )
 int item::get_gun_ups_drain() const
 {
     int draincount = 0;
-    if( type->gun.get() != nullptr ){
+    if( type->gun ){
```
Comment (Coolthulhu) [inline src/item.cpp]: Code quality regression - casting to `bool` instead of comparison.

**Comment by Coolthulhu** [comment]
`can_use` should directly mention what level is required. Having to go to item description on failed mod installation is a bother.

**Comment by mugling** [comment]
> Minor bug, probably not caused here: setting gun mode to bayonet and removing the bayonet will keep displaying the mode as bayonet until different mode is selected. Doesn't allow any exploits, though.

Can you open an issue for this

**Pull request merged**